### DISCUSSION TITLE
LDAP sync and subgroup hierarchy

### DISCUSSION THREAD

**Discussion by cod3monk** [keycloak/keycloak#33301] [Q&A]
We are using LDAP federation to sync groups using the `group-ldap-mapper`. It works quite well while we only had a few groups with a flat hierarchy, but now we want to move to a more fine grained hierarchical group model, which should be mirrored between LDAP and Keycloak.

I would like to have a group hierarchy similar to this:
```
/one/bar
/two/bar
```
note the two subgroups of the same name, which are distinct groups on Keycloak (according to their UUID and in a test realm without LDAP sync).

Once synced to LDAP they are maped into three (not four) groups in LDAP:
`cn=one,ou=groups,dc=test,dc=com` with member `cn=bar,ou=groups,dc=test,dc=com`
`cn=two,ou=groups,dc=test,dc=com` with member `cn=bar,ou=groups,dc=test,dc=com`
`cn=bar,ou=groups,dc=test,dc=com` with no members

This is not my intention as I would like to have `/one/bar` and `/two/bar` to be distinct groups and idealy show up as subgroups in LDAP (e.g., `cn=bar,cn=one,ou=groups,dc=test,dc=com`).

If I go the other way, create the hierarchy in the LDAP directory, with subgroups as intended, only the base groups are synted to Keycloak and any subgroups are skipped.

Is there a way to get the hierarchy mirroed between LDAP and Keycloak to avoid subgroup names from overlapping?

The `group-ldap-mapper` has the following configuration:

> LDAP Groups DN: `ou=groups,dc=test,dc=com`
> Group Name LDAP Attribute: `cn`
> Group Object Classes: `groupOfUniqueNames`
> Preserve Group Inheritance: on
> Ignore Missing Groups: off
> Membership LDAP Attribute: `uniqueMember`
> Membership Attribute Type: `DN`
> Membership User LDAP Attribute: `uid`
> LDAP Filter: ` `
> Mode: `LDAP_ONLY`
> User Groups Retrieve Strategy: `LOAD_GROUPS_BY_MEMBER_ATTRIBUTE`
> Member-Of LDAP Attribute: `memberOf`
> Mapped Group Attributes: ` `
> Drop non-existing groups during sync: off
> Groups Path: `/`

**Comment by Benjamin-Wiederschein** [comment]
Have you found any solution for your problem? We are facing the same issue, not being able to resolve this hierarchy issue, would make using keycloak not interesting..
